flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java

```diff
@@ -734,8 +734,14 @@ private void clearState() {
 		} catch (Exception e) {
 			onFatalError(new ResourceManagerException("Could not properly clear the job leader id service.", e));
 		}
+		clearState();
 	}
 
+	/**
+	 * Callback to clear state on leadership revocation.
+	 */
+	protected void clearState() {}
```

Comment:
We could also make the interface asynchronous and let the `ResourceManager` handle the waiting in grant leadership.

Reply:
True, I would like this a bit more, because then we don't duplicate this logic over multiple different `ResourceManager` implementations if needed.